plugins/modules/eth_call.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/python
+
+# Copyright (C) 2022, Gauthier Leonard
+# See the file LICENSE for licensing terms.
+from __future__ import absolute_import, division, print_function
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+
+import re
+from web3 import Web3
+from hexbytes import HexBytes
+
+__metaclass__ = type
+
+
+# Cast bytes and int parameters to make sure they are in the right format
+def cast_parameters(result):
+    param_types = re.match(r'.+\((.+)\)', result['function_sig']).group(1).split(',')
+
+    if len(param_types) != len(result['parameters']):
+        result['msg'] = (
+            'The number of parameters does not match function signature. '
+            f'Expected {len(param_types)}, got {len(result["parameters"])}'
+        )
+        result['failed'] = True
+        return
+
+    casted_params = []
+
+    for i, p in enumerate(result['parameters']):
+        if 'bytes' in param_types[i]:
+            casted_params.append(HexBytes(p))
+        elif 'int' in param_types[i]:
+            casted_params.append(int(p))
+        else:
+            casted_params.append(p)
+
+    result['casted_params'] = casted_params
+
+
+def eth_call(result):
+    c_chain = Web3(Web3.HTTPProvider(result['rpc_url']))
+
+    ash_router = c_chain.eth.contract(
+        address=result['contract_addr'], abi=result['abi']
+    )
+
+    result['call_return'] = ash_router.get_function_by_signature(
+        result['function_sig']
+    )(*result['casted_params']).call()
+
+
+def main():
```

Comment:
Not really a pertinent argument, but why do you use old module syntax ? (declaring main function everything) and not the new syntax based on classes ?

Reply:
The example from Ansible documentation uses this syntax: [Creating a module](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#creating-a-module). Do you have an example using classes?

Reply:
Sorry, I thought Openstack ansible modules were following a clean Ansible way, but it was not the case. What you are doing is still being used in ansible core nowadays.

For reference: https://github.com/openstack/ansible-collections-openstack/blob/master/plugins/modules/floating_ip.py